### PR TITLE
0.8 get body

### LINKED ISSUE #1338: plain-text mail with html-file attached fails to display
Current master branch.
If you compose a message and attach a html-file, the sent message will later not be properly displayed in thread mode:
Instead of the (plaintext) message text, I get

```
    This message contains a text/html part that was not
    rendered due to a missing mailcap entry. Please refer to
    item 5 in our FAQ: http://alot.rtfd.io/en/latest/faq.html
```

The sent mail is `Content-Type: multipart/mixed;` as expected. It contains a part with header:

```
Content-Type: text/html; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: quoted-printable
Content-Disposition: attachment; filename="int-pg.html"; maxlinelen="78"
```

also as expected.

### PR DESCRIPTION
**Pull request by pazz**
this simplifies our homebrew "extract_body" method in favor of the stdlibs email.message.get_body() 

It also fixes #1338 for me.

### CODE REVIEW

**Comment by pazz** [comment]
I believe that the tests are broken: Instead of individual MIME multipart(s) etc, we should really construct an email.message.Message here (which has the required `get_body()` method since 3.6.

File: alot/db/utils.py [hunk 1]
```diff
@@ -453,63 +453,34 @@ def remove_cte(part, as_string=False):
...
-            if rendered_payload:  # handler had output
-                body_parts.append(string_sanitize(rendered_payload))
-            # mark as attachment
-            elif cd:
-                part.replace_header('Content-Disposition', 'attachment; ' + cd)
-            else:
-                part.add_header('Content-Disposition', 'attachment;')
-    return u'\n\n'.join(body_parts)
+    if settings.get('prefer_plaintext'):
+        preferencelist = ('plain',)
```
Comment (lucc) [inline alot/db/utils.py]: What happens when the email only has a html part? (Those really do exists.)  As this is called "preference" I would alot expect to render the not preferred version if the preferred is not available.

**Comment by pazz** [comment]
Very good question, I don't know what [email.message's get_body](https://docs.python.org/3.6/library/email.message.html?highlight=email#email.message.EmailMessage.get_body) does in this case.
As far as I understand the docs, such (malformed!, as `text/html` should be wrapped in a `multipart/alternative` acoompanied by a `text/plain`) messages would return `None`. This should be detected and handled in our code.

Perhaps we should use preferencelist `('plain', 'html')` here (if `prefer_plaintext` is set).
An alternative would be to replace this config setting with one that lets the user directly specify the preferredlist that is passed on.

**Comment by lucc** [comment]
Yes I think we should use `('plain', 'html')` when `prefere_plaintext` is set. Especially because I can find many mails that are `text/html` but do not wrap that in a `multipart/alternative` but only have one part. I found them like this:
```sh
notmuch search --output=files mimetype:text/html | \
    while read file; do
        sed -n '/^$/q; /text.html/{ s#^#'"$file"': #; p; q; }' "$file"
    done
```

**Comment by pazz** [comment]
magic! I don't see how your sed code checks for the absense of multiparts but it works. nice.
And yes, i also received lots of mails like that.
What do you think of changing the config option to a list?

I think this should not be part of the upcoming 0.8 release. I plan to make the release very soon btw. If you urgently want something in, let me know.

**Comment by lucc** [comment]
I am living on master so I don't mind the release.

Upgrading the config option to a list could work. It is also nice because theoretically a mail could have any types in a `multipart/alternative`. I think it is just the common case that MUAs generate `text/html` + `text/plain` because people use email to send text. With a list the user has ultimate control. The only downside is that the config gets more complex (we can check typos in the boolean option but not in the list for example).

The sed code exits on the first empty line (`/^$/q`) which is the end of the email header. So if it did not find a text/html until then nothing is printed for this file. What is left are files where `text/html` is specified in the email header instead of a multipart header.

**Comment by pazz** [comment]
Quoting Lucas Hoffmann (2018-12-10 11:23:56)
> I am living on master so I don't mind the release.

cool.
 
> Upgrading the config option to a list could work. It is also nice because
> theoretically a mail could have any types in a multipart/alternative. I think
> it is just the common case that MUAs generate text/html + text/plain because
> people use email to send text. With a list the user has ultimate control. The
> only downside is that the config gets more complex (we can check typos in the
> boolean option but not in the list for example).

We could easily implement checks with configobj tough it'd make some extra work: We already do this, e.g. for theming attributes. Check alot/utils/configobj.py and its use in the settings manager..

> The sed code exits on the first empty line (/^$/q) which is the end of the
> email header. So if it did not find a text/html until then nothing is printed
> for this file. What is left are files where text/html is specified in the email
> header instead of a multipart header.

Very nice, thanks for the explanation!

**Comment by lucc** [comment]
About checking the list my argument goes like this:
1. there are many different mime types and we can't/don't want to keep a list of them
2. any mime type might be given by the sending MUA
3. our user might have custom mime type handling entries for `image/jpeg` or even `what/ever` and `never/imagined`
4. If  our user can specify a list of preferred mime types they might want to specify their special entries somewhere in the list.
5. Hence we can not check/verify the entries in the list (although we can always implement some heuristics but then we have to draw a border line and decide if we reject everything else or only warn about strange mime types in the list)
6. My conclusion is that we *should* not check the list.

**Comment by pazz** [comment]
Oh, you've overlooked that the `preferredlist` option to emails `get_body` method
has to be a list containing only entries `related`, `html`, and `plain`, with the obvious interpretation of html and plain as the corresponding mime-parts. No other options are valid.

File: alot/db/message.py [hunk 2]
```diff
@@ -268,17 +268,19 @@ def accumulate_body(self):
...
-        bodytext = extract_body(eml)
+        bodytext = self.get_text_content()
 
         # check if extracted body is empty but msg contains html parts
-        if (not bodytext and
-           'text/html' in (part.get_content_type() for part in eml.walk())):
-            return MISSING_HTML_MSG
+        eml = self.get_email()
+        if eml is not None:
+            bodypart = eml.get_body()
```
[the remark this answers is not in the document]
Comment (pazz) [reply alot/db/message.py]: I don't get it. its used in the next two lines.

**Comment by pazz** [comment]
sorry, I accidentally removed the branch when I pushed my changes.. see #1421 for the updated version.

**Pull request closed without merging**